File: quick_sort.py
"""
Simple python quick-sort
"""

def swap(a: list, b: int, c: int):
    " Swap elements of a list"
    temp = a[b]
    a[b] = a[c]
    a[c] = temp


def quickSort(a: list , array_length: int):
    """
    array: list of numbers
    array_length: len(array)
    """
    low_index = 0
    high_index = array_length - 1

    if (low_index < high_index):

        # Partition
        i = low_index
        j = high_index
        pivot = a[i]
        print("pivot: ", pivot)

        while(i < j):
            while(a[i] <= pivot and i < high_index):
                i = i+1
            while(a[j] > pivot and j > low_index):
                j = j-1
            if (i<j):
                swap(a, i, j)
            swap(a, low_index, j)
            print("array: ", a)
            print("i: ", i)
            print("j: ", j)


        # Recursive Call
        a[low_index: j] = quickSort(a[low_index: j], len(a[low_index:j]))
        a[j+1: high_index+1] = quickSort(a[j+1: high_index+1], len(a[j+1: high_index+1]))
    return a


# main
array = [12, 2, 4, 9, 1, 8]
quickSort(array, len(array))
print(array)

File: quick_sort.c
#include<stdio.h>

void qsort2(int *a, int n);
void printArray(int arr[], int size);
void swap(int *a, int *b);
int *partition(int *low, int *high);

int main () 
{
    int a[] = {100, 20, 4, 9, 0, 3, 10, 5};
    int n = sizeof(a) / sizeof(a[0]);

    printf("Original: ");
    printArray(a,n);
    qsort2(a, n); 
    printf("Sorted: ");
    printArray(a,n);
}

int *partition(int *low, int *high){

    int *i = low;
    int *j = high;
    int pivot = *i;

    while (i < j){
        while (*i <= pivot && i < high){
            i++;
        }
        while(*j > pivot && j > low) {
            j--;
        }
        if (i < j){
            swap(i, j);
        }
    }
    swap(low, j);
    return j;
}

void qsort2(int *a, int n) 
{
    int *low = a;
    int *high = a + n - 1;
    int size = high - a;

    if (low < high){
        int *j = partition(low, high);
        qsort2(a, j-a);
        qsort2(j+1, high-j); 
    }
}

void printArray(int arr[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

